### PR TITLE
Prefetch training data by multiprocessiterator

### PR DESCRIPTION
**Pull request by jnishi**
Replacing #323 

Loading kaldi's scp file is too slow, but this can be done in parallel with forward and backward calculation of other data. This can be easily done by using MultiprocessIterator.
This PR improves both of Chainer and Pytorch backends.

Though  this PR requires chainer/chainer#4972 to be merged, I add patch for current version of Chainer.

### CODE REVIEW

**Comment by jnishi** [comment]
This PR is a part of https://github.com/espnet/espnet/issues/334.

File: tools/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-chainer>=3.0.0
+chainer==4.3.1
```
Comment (sw005320) [inline tools/requirements.txt]: better to reflect this in `README.md`

Comment (jnishi) [reply tools/requirements.txt]: Thanks for comment. I fixed README.

**Comment by sw005320** [comment]
Thanks!
Can you also implement this for `tts_pytorch.py`. 
Also reflect #343 after it is merged, which is related to your PR.

**Comment by jnishi** [comment]
Fixed to resolve conflicts, and support TTS.
I checked 

```
Traceback (most recent call last):
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/chainer/training/trainer.py", line 306, in run
    update()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 149, in update
    self.update_core()
  File "/home/christopher/gitrepos/espnet-1/src/asr/asr_pytorch.py", line 119, in update_core
    loss.backward(loss.new_ones(self.ngpu))  # Backprop
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/tensor.py", line 93, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/autograd/__init__.py", line 90, in backward
    allow_unreachable=True)  # allow_unreachable flag
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/autograd/function.py", line 76, in apply
    return self._forward_cls.backward(self, *args)
  File "build/bdist.linux-x86_64/egg/warpctc_pytorch/__init__.py", line 51, in backward
    return ctx.grads * grad_output.type_as(ctx.grads), None, None, None, None, None, None
Will finalize trainer extensions and updater before reraising the exception.
^[[JTraceback (most recent call last):
  File "/home/christopher/gitrepos/espnet-1/egs/librispeech/asr1/../../../src/bin/asr_train.py", line 193, in <module>
    main()
  File "/home/christopher/gitrepos/espnet-1/egs/librispeech/asr1/../../../src/bin/asr_train.py", line 187, in main
    train(args)
  File "/home/christopher/gitrepos/espnet-1/src/asr/asr_pytorch.py", line 364, in train
    trainer.run()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/chainer/training/trainer.py", line 320, in run
    six.reraise(*sys.exc_info())
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/chainer/training/trainer.py", line 306, in run
    update()
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 149, in update
    self.update_core()
  File "/home/christopher/gitrepos/espnet-1/src/asr/asr_pytorch.py", line 119, in update_core
    loss.backward(loss.new_ones(self.ngpu))  # Backprop
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/tensor.py", line 93, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/autograd/__init__.py", line 90, in backward
    allow_unreachable=True)  # allow_unreachable flag
  File "/home/christopher/gitrepos/espnet-1/tools/venv/lib/python2.7/site-packages/torch/autograd/function.py", line 76, in apply
    return self._forward_cls.backward(self, *args)
  File "build/bdist.linux-x86_64/egg/warpctc_pytorch/__init__.py", line 51, in backward

RuntimeError: arguments are located on different GPUs at /pytorch/aten/src/THC/generated/../generic/THCTensorMathPointwise.cu:314
```

**Comment by sw005320** [comment]
@kan-bayashi, can you quickly review it? This PR still has some problem in multiple GPU pytorch ASR training, but the others seem to be OK, and I'm thinking of merging it to the dev branch for our further testing/development. Also, do you have any ideas of the above error?

**Comment by kan-bayashi** [comment]
LGTM. 
But I have no idea about above error.
I will investigate.

**Comment by sw005320** [comment]
I'm thinking of merging this PR. We can figure out the error during the development of the other items. @kan-bayashi do you agree?

**Comment by kan-bayashi** [comment]
I tried to run locally, but never get such kind of error.
Maybe warpctc version is related?

**Comment by kan-bayashi** [comment]
@sw005320 OK.

**Comment by kan-bayashi** [comment]
Hi @jnishi, I found that the error in multi-gpu pytorch was not happened with origin pytorch warpctc.
But if I use your modified version (`follow-upstream` branch), the error was occurred.
Could you investigate the difference?

**Comment by kan-bayashi** [comment]
I think this multi-gpu error is not related to prefetcher.
Maybe bug of warp-ctc.

**Comment by kan-bayashi** [comment]
@jnishi I found the solution. Please check your fork’s issue.

**Comment by jnishi** [comment]
@kan-bayashi Thanks! I can run multiple GPU training by your modification.
I have already fixed `follow-upstream`.

**Comment by sw005320** [comment]
@jnishi, I'm checking the top command and found that the current number of total processes is 5. Do you know how this number comes from? (one for main program, 2 for training data prefetch, and 2 for validation data prefetch?). I'm asking it because it consumes a lot of CPU memory (e.g., librispeech: 4GB x 5 = 20GB), and it would be better if we find a more efficient way. Probably we may not require to have 2 processes? Also, we may not need to use prefetch for the validation data? Can you test that condition?

**Comment by jnishi** [comment]
@sw005320 You are right. 2 processes are from iterators of validation dataset. If memory usage are problem, I will try to use `SequenceIterator` for validation dataset.

**Comment by sw005320** [comment]
I think it would affect to many people which uses a cluster. I'm now testing it with `n_processes=1` for the training data and `n_processes=0` for the validation data (I'm not sure `n_processes=0` would work or not, but anyway). Then, we'll only have totally 2 processes (4GB x 2 = 8GB would be acceptable). With this change, we also have to change the all `cmd.sh` in `egs` directories as follows
```
cuda_cmd="queue.pl --mem 4G --num-threads 2 --gpu 1 --config conf/gpu.conf"
```

**Comment by hirofumi0810** [comment]
@sw005320 
I found the same memory issues in CLSP.
Also, I found memory and thread options are canceled in gpu.conf.

```
option mem=* -l mem_free=$0,ram_free=$0
option mem=0          # Do not add anything to qsub_opts
option num_threads=* -pe smp $0
option num_threads=1  # Do not add anything to qsub_opts
```

**Pull request merged**